x/bandwidth/manager/manager.go

```diff
@@ -208,6 +273,262 @@ func (bm BandwidthManager) DecayMaxMPS(ctx sdk.Context) sdk.Error {
+	info, err := bm.storage.GetAppBandwidthInfo(ctx, accKey)
+	if err != nil {
+		return err
+	}
+
+	numMsgs := sdk.NewDec(info.MessagesInCurBlock)
+	// add back pre-check consumed bandwidth credit
+	info.CurBandwidthCredit = info.CurBandwidthCredit.Add(numMsgs.Mul(u))
+	// consume credit, currently allow the credit be negative
+	costPerMsg := bm.GetBandwidthCostPerMsg(ctx, u, p)
+	info.CurBandwidthCredit = info.CurBandwidthCredit.Sub(numMsgs.Mul(costPerMsg))
+	info.MessagesInCurBlock = 0
+	if err := bm.storage.SetAppBandwidthInfo(ctx, accKey, info); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (bm BandwidthManager) ReCalculateAppBandwidthInfo(ctx sdk.Context) sdk.Error {
+	totalAppStakeCoin := linotypes.NewCoinFromInt64(0)
+	// calculate all app total stake
+	for _, app := range bm.dm.GetLiveDevelopers(ctx) {
+		appStakeCoin, err := bm.vm.GetLinoStake(ctx, app.Username)
+		if err != nil {
+			return err
+		}
+		totalAppStakeCoin = totalAppStakeCoin.Plus(appStakeCoin)
+	}
+
+	if !totalAppStakeCoin.IsPositive() {
+		return nil
+	}
+
+	// calculate all app MPS quota
+	appMsgQuota, err := bm.getAppMsgQuota(ctx)
+	if err != nil {
+		return err
+	}
+
+	// calculate each app's share and expected MPS
+	params, err := bm.paramHolder.GetBandwidthParam(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, app := range bm.dm.GetLiveDevelopers(ctx) {
+		appStakeCoin, err := bm.vm.GetLinoStake(ctx, app.Username)
+		if err != nil {
+			return err
+		}
+		appStakePct := appStakeCoin.ToDec().Quo(totalAppStakeCoin.ToDec())
+		expectedMPS := appMsgQuota.Mul(appStakePct)
+		maxBandwidthCredit := expectedMPS.Mul(params.AppBandwidthPoolSize)
+
+		// first time app, refill it's current pool to the full
+		if !bm.storage.DoesAppBandwidthInfoExist(ctx, app.Username) {
+			newAppInfo := model.AppBandwidthInfo{
+				Username:           app.Username,
+				MaxBandwidthCredit: maxBandwidthCredit,
+				CurBandwidthCredit: maxBandwidthCredit,
+				MessagesInCurBlock: 0,
+				ExpectedMPS:        expectedMPS,
+				LastRefilledAt:     ctx.BlockHeader().Time.Unix(),
+			}
+
+			if err := bm.storage.SetAppBandwidthInfo(ctx, app.Username, &newAppInfo); err != nil {
+				return err
+			}
+
+		} else {
+			// refill the bandwidth credit with old expectedMPS (refill rate)
+			if err := bm.RefillAppBandwidthCredit(ctx, app.Username); err != nil {
+				return err
+			}
+			// update it's max and expected MPS
+			info, err := bm.storage.GetAppBandwidthInfo(ctx, app.Username)
+			if err != nil {
+				return err
+			}
+			info.ExpectedMPS = expectedMPS
+			info.MaxBandwidthCredit = maxBandwidthCredit
+
+			if err := bm.storage.SetAppBandwidthInfo(ctx, app.Username, info); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (bm BandwidthManager) CheckBandwidth(ctx sdk.Context, accKey linotypes.AccountKey, fee auth.StdFee) sdk.Error {
+	if bm.dm.DoesDeveloperExist(ctx, accKey) {
```

Comment:
Does affiliate account return true from this call? @Stumble

Reply:
nope